nautobot/docs/development/best-practices.md

```diff
@@ -63,3 +63,193 @@ class ExampleModel(PrimaryModel):
     slug = AutoSlugField(populate_from='name')
 ```
+
+## Filtering Models with FilterSets
+
+The following best practices must be considered when establishing new `FilterSet` classes for model classes.
+
+### Mapping Model Fields to Filters
+
+- Filtersets **must** inherit from `nautobot.extras.filters.NautobotFilterSet` (which inherits from `nautobot.utilities.filters.BaseFilterSet`)
+    - This affords that automatically generated lookup expressions (`ic`, `nic`, `iew`, `niew`, etc.) are always included
+    - This also asserts that the correct underlying `Form` class that maps the generated form field types and widgets will be included
+- FIltersets **must** publish all model fields from a model, including related fields.
+    - All fields should be provided using `Meta.fields = "__all__"` and this would be preferable for the first and common case as it requires the least maintanence and overhead and asserts parity between the model fields and the filterset filters.
+    - In some cases simply excluding certain fields would be the next most preferable e.g. `Meta.exclude = ["unwanted_field", "other_unwanted_field"]`
+    - Finally, the last resort should be explicitly declaring the desired fields using `Meta.fields =`. This should be avoided because it incurs the highest technical debt in maintaining alignment between model fields and filters.
+- In the event that fields do need to be customized to extend lookup expressions, a [dictionary of field names mapped to a list of lookups](https://django-filter.readthedocs.io/en/stable/ref/filterset.html#declaring-filterable-fields) **may** be used, however, this pattern is only compatible with explicitly declaring all fields, which should also be avoided for the common case. For example:
+
+```python
+class UserFilter(NautobotFilterSet):
+    class Meta:
+        model = User
+        fields = {
+            'username': ['exact', 'contains'],
+            'last_login': ['exact', 'year__gt'],
+        }
+```
+
+- It is acceptable that default filter mappings **may** need to be overridden with custom filter declarations, but [`filter_overrides`](https://django-filter.readthedocs.io/en/stable/ref/filterset.html#customise-filter-generation-with-filter-overrides) (see below) should be used as a first resort.
+- Custom filter definitions **must not** shadow the name of an existing model field if it is also changing the type.
+    - For example `DeviceFilterSet.interfaces` is a `BooleanFilter` that is shadowing the `Device.interfaces` related manager. This introduces problems with automatic introspection of the filterset and this pattern MUST be avoided.
+- Custom filter definitions **may** shadow the name of an existing field IF this is to adapt a filter for a human-readable field value (such as `slug`) on a related field vs. the `pk` (the default for most related fields). This pattern still relies on accessing the related field by name and then traversing into the relationship using a nested lookup.
```

Comment:
Should we take this a step further and assert the existing pattern more forcefully? Something like:

> For foreign-key related fields, the field **should** be shadowed, replacing the PK filter with a lookup based on a more human-readable value (typically `slug`, if available). A PK-based filter **should** be made available as well, generally with a name suffixed by `_id`. For example:

```python
    provider = django_filters.ModelMultipleChoiceFilter(
        field_name="provider__slug",
        queryset=Provider.objects.all(),
        to_field_name="slug",
        label="Provider (slug)",
    )
     provider_id = django_filters.ModelMultipleChoiceFilter(
        queryset=Provider.objects.all(),
        label="Provider (ID)",
    )
```

Reply:
This is tricky because it becomes hard to automate filter generation and/or widgets for this pattern unless we have some other hint to which slug value, like it would automatically always look for the slug on a related field?

Otherwise, by default, `provider` would be a `pk` (UUID) field, whereas if we thought about how we might implement automatic nested lookup expressions, `provider__slug` and  `provider__id` could both "just work". Perhaps we should keep the "should" and the "how" delineated but I feel like this is the time to suss these things out.

Reply:
Okay unless we go the distance and also implement the primitives I've suggested like our own `SlugPkMultipleChoiceFilter` like some of my prior work:

https://github.com/jathanism/nsot/blob/cecfda0590f2f83f3f679ba415a1d8bdad091238/nsot/api/serializers.py#L85-L139

Then we have to stick with the status quo as well. So is this the right time to assert that this yet-to-be-implemented-but-we-like-the-idea pattern should be used here, and then use that to justify spending the time to implement it?

Reply:
That's pretty much the question, yes. If there's a new direction we want to go, we can at least insist that any *new* FilterSets (such as Location, say?) follow the new direction rather than propagating an existing bad pattern any further. Can even add a note if you like to the effect of "We know most existing FilterSets aren't following this pattern, but we plan to change that in a major release".